Userland/Libraries/LibWeb/Painting/PaintingCommandExecutorCPU.cpp

```diff
@@ -60,129 +78,138 @@ CommandResult PaintingCommandExecutorCPU::draw_scaled_bitmap(Gfx::IntRect const&
+    auto try_get_scaled_destination_bitmap = [&]() -> ErrorOr<NonnullRefPtr<Gfx::Bitmap>> {
+        Gfx::IntRect actual_destination_rect;
+        auto bitmap = TRY(current_painter.get_region_bitmap(destination_rect, Gfx::BitmapFormat::BGRA8888, actual_destination_rect));
+        // get_region_bitmap() may clip to a smaller region if the requested rect goes outside the painter, so we need to account for that.
+        destination_clipped_fixup = Gfx::FloatPoint { destination_rect.location() - actual_destination_rect.location() };
+        destination_rect = actual_destination_rect;
+        if (source_rect.size() != transformed_destination_rect.size()) {
+            auto sx = static_cast<float>(source_rect.width()) / transformed_destination_rect.width();
+            auto sy = static_cast<float>(source_rect.height()) / transformed_destination_rect.height();
+            bitmap = TRY(bitmap->scaled(sx, sy));
+            destination_clipped_fixup.scale_by(sx, sy);
+        }
+        return bitmap;
+    };
+
+    auto bitmap_or_error = try_get_scaled_destination_bitmap();
+    if (bitmap_or_error.is_error()) {
+        // NOTE: If the creation of the bitmap fails, we need to skip all painting commands that belong to this stacking context.
+        //       We don't interrupt the execution of painting commands because get_region_bitmap() returns an error if the requested
+        //       region is outside of the viewport (mmap fails to allocate a zero-size region), which means we can safely proceed
+        //       with execution of commands outside of this stacking context.
+        // FIXME: Change the get_region_bitmap() API to return ErrorOr<Optional<Bitmap>> and exit the execution of commands here
+        //        if we run out of memory.
+        return CommandResult::SkipStackingContext;
+    }
 
+    auto bitmap = bitmap_or_error.release_value();
     stacking_contexts.append(StackingContext {
-        .painter = stacking_context_painter,
-        .destination = {},
-        .opacity = 1,
-    });
+        .painter = AK::make<Gfx::Painter>(bitmap),
+        .opacity = opacity,
+        .destination = destination_rect,
+        .scaling_mode = CSS::to_gfx_scaling_mode(image_rendering, destination_rect, destination_rect) });
+    painter().translate(-bounding_rect.location() + destination_clipped_fixup.to_type<int>());
 
     return CommandResult::Continue;
 }
 
-CommandResult PaintingCommandExecutorCPU::pop_stacking_context_with_mask(Gfx::IntRect const& paint_rect, RefPtr<Gfx::Bitmap> const& mask_bitmap, Gfx::Bitmap::MaskKind mask_kind, float opacity)
+CommandResult PaintingCommandExecutorCPU::pop_stacking_context()
 {
     auto stacking_context = stacking_contexts.take_last();
-    auto bitmap = stacking_context.painter.target();
-    if (mask_bitmap)
-        bitmap->apply_mask(*mask_bitmap, mask_kind);
-    painter().blit(paint_rect.location(), *bitmap, bitmap->rect(), opacity);
+    if (!stacking_context.painter.is_owned()) {
```

Comment:
Without knowing all the context behind it - this feels like a very weird heuristic. `MaybeOwned` is intended for applications that don't need or want to care about whether they own something or not. Adding different branches depending on whether something is owned or not defeats the purpose, even more so if the behavior differs significantly.

Reply:
In this case it's used as an optimisation where the stacking context takes a reference to the previous stacking context's painter, rather than creates its own painter. In that case it does not need to blit the bitmap in places, because it's already done so (by directly using the painter).